Added option to download zenith delays virtually

sssangha opened the pull request:
Added end-to-end option to download/virtually access zenith delay information from the UNR Geodetic Laboratory and derive/plot statistics using the ```raiderStats.py``` program: 
1. To access zenith delay information, run ```raiderDownloadGNSS.py``` and refer to child scripts ```downloadGNSSDelays.py``` and ```getStationDelays.py```. By default, URLs from the UNR Geodetic Laboratory repository are passed to CSV named ```gnssStationList_overbbox_withpaths.csv```, before zenith delay information is accessed virtually and appended to series of CSVs deposited under ```GPS_delays``` subdirectory and then in turn appended to a master CSV entitled ```CombinedGPS_ztd.csv``` that records all zenith delay information for a specified time. This latter CSV may be directly passed to ```raiderStats.py``` to derive statistics for zenith delay information, or alternatively ```gnssStationList_overbbox.csv``` alone may be passed to just display distribution of GNSS stations. 

2. ```—download``` flag must be specified to physically download data, which is no longer necessary as data may be accessed virtually from the UNR Geodetic Laboratory. 

3. Input list of stations supported with ``` --station_file``` flag. In case both the bounding box and list of stations are specified, stations which fall outside the bounding box are printed in a warning message. 


Series of up updates made to ```statsPlot.py``` program: 
1. Gray background plotted behind grid-cells in case gridded arrays are plotted, and clean white backgrounds plotted behind stations in case just stations are plotted. 

2. Following issue #46, help menu/parser better organized. 

3. “PlateCarree” projection passed consistently in all map plot calls to address issues with irregular grid-cells in plots. 

4. Units of column for which statistics will be performed may now be specified in the commandline with argument –unit. 


Example end-to-end subroutine: 
#Call to download all data acquired at midnight UTC over CA spanning between 2016-2018 ```raiderDownloadGNSS.py -y '2016,2018' -b '38.1 38.3 237.4 237.6' -v --out products --returntime 00:00:00'``` 

#Call to plot distribution of stations and gridded station mean delay in a specific time interval: ```raiderStats.py -f products/CombinedGPS_ztd.csv -grid_delay_mean -station_distribution -ti '2016-01-01 2018-01-01'``` 


Changes yet to be implemented: 
1. Add option to explicitly retrieve spatial information for quick station plotting vs also retrieve zenith delay information. Currently the program forces the download of both. 

2. Parallelization for building CSV from remote data and plotting. Advice here in particular appreciated. @dbekaert, I wish to implement something along the lines of what is done to sort through products in the ARIA-tools ```ARIAProduct.py``` script 

3. Allow for shapefile input for spatial subset for download and/or plotting

jlmaurer commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    return statID, lat, lon
+
+
+def downloadTropoDelays(stats, years, writeDir = '.', download = False, verbose = False):
```
Hey @sssangha can you modify the codes so that the main functions are at the top of the file, with helpers such as getID, fix_lons, etc. toward the bottom? Something like parseCMD at the top, followed by downloadTropoDelays, followed by download_UNR, etc. This way these scripts will be consistent with the updates from @Askaholic.

sssangha replied:
True, restructured accordingly.

jlmaurer commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    return p.parse_args(), p
+
+
+def parseCMD():
```
Since this function appears to be the main function for this script, perhaps we should name it something more descriptive? I think this function was originally in the version I wrote so it's my own fault for calling it this!

jlmaurer replied:
@Askaholic agreed, this is something we would like to do eventually with the other scripts as well. Perhaps next PR though. :)

jlmaurer commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+           delay, sig, east_grad, north_grad = [np.full((288,), np.nan)]*4 
+           delay[~mask] = d
+           sig[~mask] = Sig
+           east_grad[~mask] = egrad
```
We may want to add a note in the file as per Geoff Blewitt's comment that they have the north gradient and east gradient swapped in their files. From his email: 

"Thanks David.   At least we don't have any data loss, which is the most important thing.

Before I forget to mention it, please note that, due to a formatting error, our tropospheric gradient columns are interchanged.    That is, the column labeled as the East gradient is actually the North, and vice versa.   But you may be only using ZTD in which case, no problem."

If in the future anyone wants to use the gradients we could leave a comment directing them to check with the UNR folks about the proper format.

dbekaert replied:
@jlmaurer @sssangha perhaps a broader question would be do we even need to gradient in east, north, up. We leverage the delays themselves nto the gradients. This might just add more overhead to our files that is not used by RAiDER.

jlmaurer replied:
@dbekaert True but the overhead seems likely to be minimal; as long as we're doing nothing but passing the values; seems like the info it could be useful to some users, but perhaps not.

sssangha replied:
Ok, added note/reference to that formatting error and made sure to flip the order of the initialized variables to account for this. I agree with @jlmaurer that we should just pass instead of omitting altogether

jlmaurer commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+import os
+import pandas as pd
+
+def getDate(stationFile):
```
Same comment here about ordering the functions from most to least important/first to last called.

jlmaurer commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+    if not os.path.exists(pathbase):
+        os.mkdir(pathbase)
+
+    returnTime=(int(returnTime.split(':')[0])*3600)+(int(returnTime.split(':')[1])*60)+int(returnTime.split(':')[2])
```
Might be a little cleaner to call this part that parses the returnTime as its own function

Askaholic replied:
Lots of ways to do this. Definitely extract it to a function though. Probably the most straight forward is:
```suggestion
    h, m, s = returnTime.split(":")
    returnTime = int(h * 3600 + m * 60 + s)
```

sssangha replied:
Make sense, moved out to a separate function

jlmaurer commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+#!/usr/bin/env python3
```
@sssangha can you also run the following command on each of the modified files to take care of whitespaces consistent with PR#58:
autopep8 --select E231,E225,E251,E111,E128,E127,E265,E303,E261,E301 --in-place --recursive  <filename>

Askaholic commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    return pd.DataFrame(results).set_index('ID')
+            
+
+def parse_years(input):
```
Not exactly sure how the edge cases are supposed to be handled, but I feel like you can make use of the 2 argument version of `range` i.e.
```python3
assert list(range(10, 10)) == []
assert list(range(10, 11)) == [10]
assert list(range(10, 15)) == [10, 11, 12, 13, 14]
assert list(range(20, 10)) == []
```

sssangha replied:
Ok, I was thinking of taking this and leveraging map as so:
```
def parse_years(timestr):
    '''
    Takes string input and returns a list of years as integers
    '''
    years = list(map(int, timestr.split(',')))
    # If two years non-consecutive years are given, query for each year in between
    if len(years) == 2:
         years = list(range(years[0],years[1]+1))
    return years
```

In this way, inputs non-consecutive pairs of years are expanded (e.g. '1994,1997' --> [1994 1995 1996 1997]), while inputs of multiple years are left alone (e.g. '2014,2015,2016' --> [2014 2015 2016])

Askaholic replied:
Very nice!

Askaholic commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+
+    # Setup bounding box
+    if args.bounding_box:
+        if isinstance ([str(val) for val in args.bounding_box.split()], list) and not os.path.isfile(args.bounding_box):
```
I don't think it's possible for this `isinstance` check to fail. You are constructing a list, then checking if it is a list, and then throwing the list away again?

sssangha replied:
Yeah, should've clarified that this is a placeholder for an option whereby a user may pass a valid shapefile, as is done in the ARIA-tools package. Will require more work to integrate, hence why I left off without implementing it yet.

Askaholic commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+#
+import copy
```
Unused import

Askaholic commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+    '''
+
+    # find the date info
+    year = int(os.path.basename(stationFile).split('.')[1])
```
Again, split once and unpack to your variables

Askaholic commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+         (delay in mm, delay uncertainty, delay gradients)
+    '''
+    #import functions
+    import gzip, zipfile, gzip, requests, io
```
Put at top of file

Askaholic commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+        for line in f.readlines():
+            try:
+                line = line.decode('utf-8')
+            except:
```
```suggestion
            except UnicodeDecodeError:
```

Askaholic commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+                # Break iteration if time from line in file does not match date reported in filename
+                if doy != doyFromFile:
+                   print('WARNING: time %s from line in conflict with time %s from file %s, will continue reading next tarfile(s)' \
+                       %(doy,doyFromFile,j))
```
use `.format` instead of `%`

Askaholic commented on tools/RAiDER/getStationDelays.py:
```diff
@@ -0,0 +1,184 @@
+        del f
+        # Break iteration if file contains no data.
+        if d == []:
+            print('WARNING: file %s for station %s is empty, will continue reading next tarfile(s)'%(j, j.split('.')[0]))
```
same

Askaholic commented on tools/RAiDER/statsPlot.py:
```diff
@@ -400,11 +437,27 @@ def _getExtent(self):  # dataset, spacing=1, userbbox=None
         data = pd.read_csv(self.fname)
+        #check if user-specified key is valid
+        if self.col_name not in data.keys():
+            raise Exception('User-specified key %s not found in inpuit file %s. Must specify valid key.'%(self.col_name,self.fname))
```
Use `.format`. Also typo in word `input`

Askaholic commented on tools/RAiDER/statsPlot.py:
```diff
@@ -494,19 +547,24 @@ def __call__(self, gridarr, plottype, workdir='./', drawgridlines=False, colorba
+        # reset in case bounds too large
+        try:
+            axes.set_extent(self.plotbbox, ccrs.PlateCarree())
+        except:
```
Should either specify the error type or explicitly catch all exceptions with
```suggestion
        except Exception:
```

jlmaurer commented on tools/RAiDER/statsPlot.py:
```diff
@@ -180,7 +216,7 @@ def resid(x, d, v, m):
             lb = np.zeros(Nparm)
             x0 = (ub-lb)/2
-
+     
```
@Askaholic if you know a way to do this in vim you'll be my hero...

Askaholic replied:
In VIM it's probably a bit more involved than ticking a checkbox, but if you're using VIM then you know what you signed up for :P.

https://vi.stackexchange.com/questions/454/whats-the-simplest-way-to-strip-trailing-whitespace-from-all-lines-in-a-file

sssangha replied:
I use gedit. Old habits die hard!

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+
+    if userstatList:
+        with open(userstatList) as f:
+            userstatList = [line.rstrip('\n') for line in f]
```
Nope, ```rstrip``` was a vestige, and ironically the read file function was treated as one too. Will replace this with a call to the read file function instead.

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    llhBox should be a tuple with format (lat1, lat2, lon1, lon2), where lat1, lon1 define the lower left-hand corner and lat2, lon2 
+    define the upper right corner. 
+    '''
+    import pandas as pd
+    from urllib.request import urlopen 
```
Yeah, pandas was redundant here anyways.

And I don't believe there was a particular reason for choosing ```urlib``` over ```requests```. Any reason to proceed with the latter though @Askaholic?

Askaholic replied:
Mostly because `requests` is a higher level library. It will be able to handle stuff like redirects, cookies, etc out of the box. Even the documentation of `urllib` recommends using `requests` instead https://docs.python.org/3/library/urllib.request.html#module-urllib.request

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+        return url
+
+
+def download_url(url, save_path, verbose = False, chunk_size=128):
```
These are smaller files anyways, but I will set to ```2048``` and test it out to see if it makes a difference.

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    '''
+    r = requests.get(url, stream=True)
+    if r.status_code==404:
+        return ''
```
This is intentional. I want to pass empty fields to the master CSV file so we can see specifically which years data from a given station are not available.

Askaholic replied:
Ok, IMO it would be cleaner to do the error handling in the calling function. Since `download_url` seems quite generic to me.

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    r = requests.get(url, stream=True)
+    if r.status_code==404:
+        return ''
+    else:
```
True, but since I want to return empty fields in case the URL isn't valid (see my above comment) I suppose I could set the url variable to an empty string. I.e.:
``
    if r.status_code == 404:
        url = ''
    return url
``

Askaholic replied:
If you simply delete the `else:` (And dedent the contents of the Else block) the function will behave exactly the same way. It’s a nice way of saving a level of indentation, which in turn makes it easier to read.

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+    '''
+    Read a list of GNSS station names from a plain text file
+    '''
+    statList = []
+    with open(filename, 'r') as f:
+        for line in f:
+            statList.append(line.strip())
+    return statList 
```
Yeah, probably ```readTextFile``` should suffice. We should avoid ```readFile``` as that conflicts with a python utils module

sssangha commented on tools/RAiDER/downloadGNSSDelays.py:
```diff
@@ -0,0 +1,330 @@
+        bbox    - length-4 list of floats that describes a bounding box. Format is 
+                  S N W E
+    '''
+    if writeLoc is None:
+        writeLoc = os.path.join(os.getcwd(), 'gnssStationList_overbbox.csv')
+    else:
+        writeLoc = os.path.join(writeLoc, 'gnssStationList_overbbox.csv')
```
Always hit my head when I miss an opportunity for a single-liner! :p

sssangha commented on tools/RAiDER/statsPlot.py:
```diff
@@ -147,10 +180,13 @@ def _binnedVario(self, hEff, rawVario, xBin=None):
-                expVario.append(np.nanmean(rawVario[iBinMask]))
+                #supress mean of empty slice warning
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore", category=RuntimeWarning)
```
Yeah, ```warnings.filterwarnings("ignore", message="Mean of empty slice")``` suffices.

sssangha commented:
> Thanks for the contributions.
> @sssangha can you also update setup.py accordingly to install the new script?

Done in latest commit.

sssangha commented:
Ok, looks like I addressed all suggestions. As @jlmaurer suggested though, I will update the structure of the parser/main function in a separate PR, along with parallelization of downloading/plotting.
